warehouse/oidc/utils.py

```diff
@@ -14,51 +14,129 @@
         repository_owner, repository_name = repository.split("/", 1)
         workflow_prefix = f"{repository}/.github/workflows/"
         workflow_ref = signed_claims["job_workflow_ref"].removeprefix(workflow_prefix)
 
+        provider_cls = GitHubProvider if not pending else PendingGitHubProvider
+
         return (
-            session.query(GitHubProvider)
+            session.query(provider_cls)
             .filter_by(
                 repository_name=repository_name,
                 repository_owner=repository_owner,
                 repository_owner_id=signed_claims["repository_owner_id"],
             )
             .filter(
                 literal(workflow_ref).like(
-                    func.concat(GitHubProvider.workflow_filename, "%")
+                    func.concat(provider_cls.workflow_filename, "%")  # type: ignore
                 )
             )
             .one_or_none()
         )
     else:
         # Unreachable; same logic error as above.
         return None  # pragma: no cover
+
+
+def reify_pending_provider(
+    session, pending_provider: PendingOIDCProvider, remote_addr: str
+):
+    """
+    Reify a `PendingOIDCProvider` into an `OIDCProvider`, creating its
+    project in the process.
+
+    `remote_addr` is the IP address to attribute the changes to, in both the journal
+    and event logs.
+
+    Deletes the pending OIDC provider once complete.
+
+    Returns the a tuple of the new project and reified OIDC provider.
+    """
+    new_project = Project(name=pending_provider.project_name)  # type: ignore
+    session.add(new_project)
+
+    session.add(
+        JournalEntry(
+            name=new_project.name,
+            action="create",
+            submitted_by=pending_provider.added_by,  # type: ignore
+            submitted_from=remote_addr,
+        )
+    )
+
+    new_project.record_event(
+        tag=EventTag.Project.ProjectCreate,
+        ip_address=remote_addr,
+        additional={"created_by": pending_provider.added_by.username},  # type: ignore
+    )
+
+    session.add(
+        Role(
+            user=pending_provider.added_by,  # type: ignore
+            project=new_project,
+            role_name="Owner",
+        )
+    )
+
+    session.add(
+        JournalEntry(
+            name=new_project.name,
+            action=f"add Owner {pending_provider.added_by.username}",  # type: ignore
+            submitted_by=pending_provider.added_by,  # type: ignore
+            submitted_from=remote_addr,
+        )
+    )
+    new_project.record_event(
+        tag=EventTag.Project.RoleAdd,
+        ip_address=remote_addr,
+        additional={
+            "submitted_by": pending_provider.added_by.username,  # type: ignore
+            "role_name": "Owner",
+            "target_user": pending_provider.added_by.username,  # type: ignore
+        },
+    )
+
+    new_provider = pending_provider.reify(session)
+    new_project.oidc_providers.append(new_provider)  # type: ignore
```

Comment:
Per the earlier comments: I'm not 100% sure why I need so many type ignores here, given that similar code elsewhere in Warehouse checks just fine. I've added them just to get the CI passing.